includes/mailpoet/class-sensei-mailpoet.php

```diff
@@ -0,0 +1,374 @@
+		}
+		return $this->mail_poet_lists;
+	}
+
+	/**
+	 * Add students as subscribers to lists on MailPoet for courses/groups.
+	 *
+	 * @since $$next-version$$
+	 * @param array      $students A list of students belonging to this group/course.
+	 * @param string|int $list_id ID of the list on MailPoet.
+	 *
+	 * @return void
+	 */
+	public function add_subscribers( $students, $list_id ) {
+		$options = array(
+			'send_confirmation_email'      => false,
+			'schedule_welcome_email'       => false,
+			'skip_subscriber_notification' => true,
+		);
+		foreach ( $students as $student ) {
+			$subscriber_data = array(
+				'email'      => $student['email'],
+				'first_name' => $student['display_name'],
+			);
+
+			try {
+				// All WordPress users are already on a list 'WordPress Users' on MailPoet.
+				$mp_subscriber = $this->mailpoet_api->getSubscriber( $student['email'] );
+				$this->mailpoet_api->subscribeToList( $mp_subscriber['id'], $list_id, $options );
+			} catch ( \MailPoet\API\MP\v1\APIException $exception ) {
+				if ( 4 === $exception->getCode() ) {
+					// subscriber does not exist.
+					$this->mailpoet_api->addSubscriber( $subscriber_data, array( $list_id ), $options );
+				}
+			}
+		}
+	}
+
+	/**
+	 * Remove students as subscribers from lists on MailPoet for courses/groups.
+	 *
+	 * @since $$next-version$$
+	 * @param array      $subscribers A list of students to remove.
+	 * @param string|int $list_id ID of the list on MailPoet.
+	 *
+	 * @return void
+	 */
+	public function remove_subscribers( $subscribers, $list_id ) {
+		foreach ( $subscribers as $subscriber ) {
+			try {
+				$mp_subscriber = $this->mailpoet_api->getSubscriber( $subscriber['email'] );
+				$this->mailpoet_api->unsubscribeFromList( $mp_subscriber['id'], $list_id );
+			} catch ( \MailPoet\API\MP\v1\APIException $exception ) {
+				continue;
+			}
+		}
+	}
+
+	/**
+	 * Creates a Sensei LMS MailPoet list with name and description.
+	 *
+	 * @since $$next-version$$
+	 * @param string $list_name The name of the list.
+	 * @param string $list_description The description of the list.
+	 *
+	 * @return int|string|null
+	 */
+	public function create_list( $list_name, $list_description ) {
+		$new_list = array(
+			'name'        => $list_name,
+			'description' => $list_description,
+		);
+		try {
+			$new_list = $this->mailpoet_api->addList( $new_list );
+			return $new_list['id'];
+		} catch ( \MailPoet\API\MP\v1\APIException $exception ) {
+			// see https://github.com/mailpoet/mailpoet/blob/trunk/doc/api_methods/AddList.md#error-handling.
+			return null;
+		}
+	}
+
+	/**
+	 * Add a student as a subscriber to a list on MailPoet for a course or group.
+	 * If the list does not exist, it will be created.
+	 *
+	 * @param int $post_id Post ID.
+	 * @param int $user_id User ID.
+	 *
+	 * @access private
+	 * @return void
+	 */
+	public function add_student_subscriber( $post_id, $user_id ) {
+		$student = get_user_by( 'id', $user_id );
```

Comment:
Need to change the student to array (and lowercase the email and everything) here too, otherwise there is a fatal error when a student enrolls in a course, and also when adding someone to a group:

>  Fatal error: Uncaught Error: Cannot use object of type WP_User as array
in /Users/yscik/dev/sensei/includes/mailpoet/class-sensei-mailpoet.php on line 149

Call stack:

    Sensei_MailPoet::add_subscribers()
    /Users/yscik/dev/sensei/includes/mailpoet/class-sensei-mailpoet.php:234
    Sensei_MailPoet::add_student_subscriber()
    /Users/yscik/dev/sensei/includes/mailpoet/class-sensei-mailpoet.php:309
    Sensei_MailPoet::add_student_course_subscriber()
    wp-includes/class-wp-hook.php:307

Reply:
Seeing the same error when removing a student from a group. And I think it's also happening when removing enrollment from a course, because the student wasn't removed from the list.

> <b>Fatal error</b>: Uncaught Error: Cannot use object of type WP_User as array<br>in <b>/Users/yscik/dev/sensei/includes/mailpoet/class-sensei-mailpoet.php</b> on line <b>178</b>

<ol>
	<li><code>Sensei_MailPoet::remove_subscribers()</code><br><span class="qm-info qm-supplemental">sensei/includes/mailpoet/class-sensei-mailpoet.php:272</span>
	</li>
	<li><code>Sensei_MailPoet::remove_student_subscribers()</code><br><span class="qm-info qm-supplemental">wp-includes/class-wp-hook.php:307</span>
	</li>
	<li><code>WP_Hook::apply_filters()</code><br><span class="qm-info qm-supplemental">wp-includes/class-wp-hook.php:331</span>
	</li>
	<li><code>WP_Hook::do_action()</code><br><span class="qm-info qm-supplemental">wp-includes/plugin.php:476</span>
</ol>